src/librustc/middle/mem_categorization.rs

```diff
@@ -894,27 +883,7 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         debug!("cat_rvalue_node(id={:?}, span={:?}, expr_ty={:?})",
                hir_id, span, expr_ty);
 
-        let promotable = self.rvalue_promotable_map.as_ref().map(|m| m.contains(&hir_id.local_id))
-                                                            .unwrap_or(false);
-
-        debug!("cat_rvalue_node: promotable = {:?}", promotable);
-
-        // Always promote `[T; 0]` (even when e.g., borrowed mutably).
-        let promotable = match expr_ty.kind {
-            ty::Array(_, len) if len.try_eval_usize(self.tcx, self.param_env) == Some(0) => true,
-            _ => promotable,
-        };
-
-        debug!("cat_rvalue_node: promotable = {:?} (2)", promotable);
-
-        // Compute maximum lifetime of this rvalue. This is 'static if
-        // we can promote to a constant, otherwise equal to enclosing temp
-        // lifetime.
-        let re = if promotable {
-            self.tcx.lifetimes.re_static
-        } else {
-            self.temporary_scope(hir_id.local_id)
-        };
+        let re = self.temporary_scope(hir_id.local_id);
         let ret = self.cat_rvalue(hir_id, span, re, expr_ty);
```

Comment:
Can that lifetime be removed from MC? It would be wrong now, so I hope nothing uses it.

Reply:
`regionck` was using it, but it was already getting the wrong lifetimes.